README.md
# D3 Scatterplot Matrices

[![Travis-CI Build Status](https://travis-ci.org/garthtarr/pairsD3.svg?branch=master)](https://travis-ci.org/garthtarr/pairsD3) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/pairsD3)](http://cran.r-project.org/package=pairsD3) [![](http://cranlogs.r-pkg.org/badges/pairsD3)](http://cran.rstudio.com/web/packages/pairsD3/index.html)

This is a port of Mike Bostock's [D3 scatter plot matrix](http://bl.ocks.org/mbostock/4063663) code to the [htmlwidgets](https://github.com/ramnathv/htmlwidgets) framework.  There have been some minor adjustments, including the addition of tooltips.

You could also consider the [pairedVis()](https://healthvis.wordpress.com/2013/04/05/pairedvis/) function in the [healthvis](https://healthvis.wordpress.com/) package.

Take it for a test run [here](http://garthtarr.com/apps/pairsD3/) (you can upload your own data).

## Installation

The `pairsD3` package is available on CRAN:

```s
install.packages("pairsD3")
```

Alternatively, you can install the development version of `pairsD3` from Github using the `devtools` package as follows:

```s
devtools::install_github("garthtarr/pairsD3")
```

## Usage

A canonical example with the iris data:

```s
data(iris)
require(pairsD3)
pairsD3(iris[,1:4],group=iris[,5])
```

#### Save

Use `savePairs` to save a pairs plot as a stand alone HTML file:

```s
library(magrittr)
pairsD3(iris[,1:4],group=iris[,5]) %>% savePairs(file = 'iris.html')
```

#### Shiny

You can view an interactive scatterplot matrix using the `shinypairs` function:

```s
shinypairs(iris)
```

#### Rmarkdown

You can include interactive scatterplot matrices in rmarkdown documents in the usual way:

    ```{r}
    require(pairsD3)
    pairsD3(iris)
    ```

#### Slidify

HTML widgets are not (yet) supported in slidify.  A workaround is to do save the widget as a webpage then include that webpage in slidify using an iframe:


    ```{r, results='asis',echo=FALSE}
    require(pairsD3)
    pd3 = pairsD3(iris)
    savePairs(pd3, 'pD3.html')
    cat('<iframe src="pD3.html"> </iframe>')
    ```




NEWS
Version 0.1.2
-------------

* Added option for main diagonal
* Initialised the documentation page with pkgdown

Version 0.1.1
-------------

* Minor bug fixes

Version 0.1.0
-------------

* Specify d3 as a dependency and move it to its own folder.
  This is to ensure that d3 is picked up only once on a page,
  when multiple d3 based charts are used on a page. Thanks to
  Ramnath Vaidyanathan for this fix.
* Namespace styles so that they are applied only to pairsD3 charts.
  Thanks to Ramnath Vaidyanathan for this fix.
* Better handling of NA and NaN values.  These are now dealt with
  on a pairwise basis.
* Updated tooltip to show row number (name) and (optionally) the
  group an observation belongs to.
* Improved the padding defaults and sizing.

Version 0.0.6
-------------

* First CRAN release of pairsD3 package


R/pairsD3.R
#' D3 scatterplot matrices
#'
#' An interactive matrix of scatterplots is produced.
#'
#' @param x the coordinates of points given as numeric columns of a
#'   matrix or data frame. Logical and factor columns are converted
#'   to numeric in the same way that \code{data.matrix} does.
#' @param group a optional vector specifying the group each observation
#'   belongs to.  Used for tooltips and colouring the observations.
#' @param subset an optional vector specifying a subset of observations
#'   to be used for plotting. Useful when you have a large number of
#'   observations, you can specify a random subset.
#' @param labels the names of the variables (column names of \code{x}
#'   used by default).
#' @param cex the magnification of the plotting symbol (default=3)
#' @param col an optional (hex) colour for each of the levels in the group
#'   vector.
#' @param big a logical parameter.  Prevents inadvertent plotting of huge
#'   data sets.  Default limit is 10 variables, to plot more than 10 set
#'   \code{big=TRUE}.
#' @param theme a character parameter specifying whether the theme should
#'   be colour \code{colour} (default) or black and white \code{bw}.
#' @param width the width (and height) of the plot when viewed externally.
#' @param opacity numeric between 0 and 1. The opacity of the plotting
#'   symbols (default 0.9).
#' @param tooltip an optional vector with the tool tip to be displayed when
#'   hovering over an observation. You can include basic html.
#' @param leftmar space on the left margin
#' @param topmar space on the bottom margin
#' @param diag logical, whether or not the main diagonal is plotted
#'   (scatter plot of variables against themselves).
#'
#' @import htmlwidgets
#'
#' @examples
#' data(iris)
#' \dontrun{
#' pairsD3(iris[,1:4],group=iris[,5],
#'          labels=gsub(pattern = "\\.",replacement = " ", names(iris)))
#' }
#'
#' @export
pairsD3 <- function(x, group = NULL, subset = NULL, labels = NULL, cex = 3,
                    width = NULL, col = NULL, big = FALSE, theme = "colour", opacity = 0.9,
                    tooltip = NULL, leftmar = 35, topmar = 2,
                    diag = FALSE) {
  height=width
  # ensure the data is a numeric matrix but also an array
  data = data.frame(data.matrix(x))
  n = dim(data)[1]
  p = dim(data)[2]
  if(!big & dim(data)[2]>=10){
    warning("If you are sure you want that many variables plotted, set big=TRUE")
    return(NULL)
  }
  if(is.null(labels)){
    labels=names(data)
  }
  if(is.null(group)){
    group = rep("",n)
  }
  n.group = length(levels(factor(group)))
  groupval = as.numeric(factor(group))-1
  if(is.null(tooltip)){
    if(is.null(rownames(x))){
      tooltip = c(1:n)
    } else {
      tooltip = rownames(x)
    }
    if(n.group>1){
      tooltip=paste(tooltip,"<br/>",group)
    }
  }
  alldata = cbind(data,groupval,group,tooltip)
  if(is.null(col)){
    if(is.element(theme,c("colour","color"))){
      # Set1 from brewer.pal() in the RColorBrewer package
      col=c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00",
            "#FFFF33", "#A65628", "#F781BF", "#999999")[1:n.group]
      while(any(is.na(col))){
        col[is.na(col)] = col[1:sum(is.na(col))] # repeat colours
      }
    } else if(theme=="bw"){
      col=grDevices::gray.colors(n.group,start=0,end=0.75)
    }
  }
  if(length(col)>n.group){
    warning("The length of col should be the same as the number of levels in
             the groups vector.")
    col = unique(col)
  }
  # create a list that contains the settings
  settings <- list(
    width = width,
    height = height,
    col = col,
    cex = cex,
    opacity = opacity,
    diag = diag
  )
  # pass the data and settings using 'xin'
  xin <- list(
    data = data,
    group = group,
    alldata = alldata,
    n = n,
    p = p,
    labels = labels,
    settings = settings,
    leftmar = leftmar,
    topmar = topmar
  )
  # create widget
  htmlwidgets::createWidget(
    name = 'pairsD3',
    x = xin,
    width = width,
    height = height,
    sizingPolicy = htmlwidgets::sizingPolicy(padding = 0, browser.fill = TRUE),
    package = 'pairsD3'
  )
}

#' Widget output function for use in Shiny
#'
#' @param outputId Shiny output ID
#' @param width width default '100\%'
#' @param height height default '400px'
#'
#' @export
pairsD3Output <- function(outputId, width = '100%', height = '400px'){
  shinyWidgetOutput(outputId, 'pairsD3', width, height, package = 'pairsD3')
}

#' Widget render function for use in Shiny
#'
#' @param expr pairsD3 expression
#' @param env environment
#' @param quoted logical, default = FALSE
#'
#' @export
renderPairsD3 <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  shinyRenderWidget(expr, pairsD3Output, env, quoted = TRUE)
}


inst/htmlwidgets/pairsD3.js
HTMLWidgets.widget({
  name: 'pairsD3',
  type: 'output',
  initialize: function(el, width, height) {
    return {};
  },

  renderValue: function(el, xin, instance) {
    // save params for reference from resize method
    instance.xin = xin;
    // draw the graphic
    this.drawGraphic(
      el,
      xin,
      el.getBoundingClientRect().width,
      el.getBoundingClientRect().height
    );


  },

  drawGraphic: function(el, xin, width, height){
    // remove existing children
    while (el.firstChild)
      el.removeChild(el.firstChild);

    wide = HTMLWidgets.dataframeToD3(xin.data);
    factor = xin.groupval;
    alldata = HTMLWidgets.dataframeToD3(xin.alldata);
    legdata = HTMLWidgets.dataframeToD3(xin.legdata);

    domainByTrait = {}; //
    traits = d3.keys(wide[0]); // column names
    p = traits.length; // number of variables

    traits.forEach(function(trait) {
      domainByTrait[trait] = d3.extent(wide,
                              function(d) { return d[trait]; });
    });

    // get the width and height
    //var width = el.offsetWidth;
    //var height = el.offsetHeight0
    // var width = 960;
    // var size = 150;
    var xinlab = xin.labels;
    var padding = 10;
    var size = (d3.min([width,height])-2*padding)/p-4;
    var color = [];
    if(xin.settings.col.constructor===Array){
      color = xin.settings.col;
    } else {
      color = [xin.settings.col];
    }

    var x = d3.scale.linear()
            .range([padding / 2, size - padding / 2]);

    var y = d3.scale.linear()
            .range([size - padding / 2, padding / 2]);

    var xAxis = d3.svg.axis()
                .scale(x)
                .orient("bottom")
                .ticks(3);

    var yAxis = d3.svg.axis()
                .scale(y)
                .orient("left")
                .ticks(3);

    // add the tooltip area to the webpage
    var tooltip = d3.select(el).append("div")
          .attr("class", "tooltip")
          .style("opacity", 0);

    svg = d3.select(el).append("svg")
          .attr("width", size * p + padding*2 + xin.leftmar)
          .attr("height", size * p + padding*2 + xin.topmar)
          .append("g")
          .attr("transform", "translate(" + xin.leftmar + "," + xin.topmar + ")");

    xAxis.tickSize(size * p);
    yAxis.tickSize(-size * p);

    var brush = d3.svg.brush()
                .x(x)
                .y(y)
                .on("brushstart", brushstart)
                .on("brush", brushmove)
                .on("brushend", brushend);

    var brushCell;

    // Clear the previously-active brush, if any
    function brushstart(p) {
      if (brushCell !== this) {
        d3.select(brushCell).call(brush.clear());
        x.domain(domainByTrait[p.x]);
        y.domain(domainByTrait[p.y]);
        brushCell = this;
      }
    }

    // Highlight the selected circles
    function brushmove(p) {
      var e = brush.extent();
      svg.selectAll("circle").classed("greyed",
        function(d) { return e[0][0] > d[p.x] || d[p.x] > e[1][0] || e[0][1] > d[p.y] || d[p.y] > e[1][1];
        });
    }

    function brushend() {
      // If the brush is empty, select all circles.
      if (brush.empty()){
        svg.selectAll(".greyed").classed("greyed", false);
      }
      // Identify selected observations and pass them to Shiny as
      // input$selectedobs
      if(typeof Shiny !== 'undefined'){
        var circleS = svg.selectAll('circle')[0]
                        .map(function(d) {return d.className['baseVal']});
        Shiny.onInputChange("selectedobs", circleS);
      }
    }

    // X-axis
    svg.selectAll(".x.axis")
        .data(traits)
        .enter().append("g")
        .attr("class", "x axis")
        .attr("transform",
              function(d, i) { return "translate(" + i * size + ",0)"; })
        .each(function(d) { x.domain(domainByTrait[d]);
                            d3.select(this).call(xAxis); });

    // Y-axis
    svg.selectAll(".y.axis")
        .data(traits)
        .enter().append("g")
        .attr("class", "y axis")
        .attr("transform",
              function(d, i) { return "translate(0," + i * size + ")"; })
        .each(function(d) { y.domain(domainByTrait[d]);
                            d3.select(this).call(yAxis); });
    // Cell and plot
    cell = svg.selectAll(".cell")
              .data(cross(traits, traits))
              .enter().append("g")
              .attr("class", "cell")
              .attr("transform",
                    function(d) {
                      return "translate(" + d.i * size + "," + d.j * size + ")";
                    })
              .each(plot);
    // Titles for the diagonal.
    cell.filter(function(d) { return d.i === d.j; }).append('text')
      .attr("x", size/2)
      .attr("y", size/2)
      .text(function(d) { return xinlab[d.i]; })
      .style("text-anchor", "middle");

    // tooltip fn
    if(typeof Shiny !== 'undefined'){
        leftoffset = document.getElementById("pairsplot").offsetParent.offsetLeft
        topoffset = document.getElementById("pairsplot").offsetParent.offsetTop
    } else {
        leftoffset = 0
        topoffset = 0
    }

    // plot function
    function plot(p) {
      var cell = d3.select(this);
      x.domain(domainByTrait[p.x]);
      y.domain(domainByTrait[p.y]);
      // Plot frame
      cell.append("rect")
        .attr("class", "frame")
        .attr("x", padding / 2)
        .attr("y", padding / 2)
        .attr("width", size - padding)
        .attr("height", size - padding);
      // apply the brush needs to be done before tooltip
      cell.call(brush)
      // plot the data
      if(xin.settings.diag){
        cell.selectAll("circle")
          .data(alldata)
          .enter().append("circle")
          .attr("cx", function(d) { return x(d[p.x]); })
          .attr("cy", function(d) { return y(d[p.y]); })
          .attr("r", xin.settings.cex)
          .style("display", function(d) { return d[p.x]==null || d[p.y]==null ? "none" : null; })
          .style("fill", function(d) { return color[d.groupval]; })
          .style("opacity", xin.settings.opacity)
          .on("mouseover", function(d) {
            tooltip.transition()
              .duration(200)
              .style("opacity", .9);
            tooltip.html(d.tooltip)// + "<br/> (" + xValue(d) + ", " + yValue(d) + ")")
              //.placement("right")
              .style("left", (event.pageX - leftoffset + 1) + "px")
              .style("top", (event.pageY - topoffset - 20) + "px");
          })
          .on("mouseout", function(d) {
            tooltip.transition()
              .duration(500)
              .style("opacity", 0);
          });
      } else {
        if (p.x !== p.y){ // prevents a main diagonal being plotted
        cell.selectAll("circle")
          .data(alldata)
          .enter().append("circle")
          .attr("cx", function(d) { return x(d[p.x]); })
          .attr("cy", function(d) { return y(d[p.y]); })
          .attr("r", xin.settings.cex)
          .style("display", function(d) { return d[p.x]==null || d[p.y]==null ? "none" : null; })
          .style("fill", function(d) { return color[d.groupval]; })
          .style("opacity", xin.settings.opacity)
          .on("mouseover", function(d) {
            tooltip.transition()
              .duration(200)
              .style("opacity", .9);
            tooltip.html(d.tooltip)// + "<br/> (" + xValue(d) + ", " + yValue(d) + ")")
              //.placement("right")
              .style("left", (event.pageX - leftoffset + 1) + "px")
              .style("top", (event.pageY - topoffset - 20) + "px");
          })
          .on("mouseout", function(d) {
            tooltip.transition()
              .duration(500)
              .style("opacity", 0);
          });
      }
      }
    }


    // cross function
    function cross(a, b) {
      var c = [], n = a.length, m = b.length, i, j;
      for (i = -1; ++i < n;) {
        for (j = -1; ++j < m;) {
          c.push({x: a[i], i: i, y: b[j], j: j})
        }
      };
      return c;
    }

  },

  resize: function(el, width, height, instance) {
    if(instance.xin){
      this.drawGraphic(el, instance.xin, width, height);
    }
  }
});




R/shinypairs.R
#' Shiny interface to the pairsD3 function
#'
#' Opens a shiny GUI to facilitate interaction with the pairsD3 function
#'
#' @param x the coordinates of points given as numeric columns of a
#'   matrix or data frame. Logical and factor columns are converted
#'   to numeric in the same way that \code{data.matrix} does.
#' @param group a optional vector specifying the group each observation
#'   belongs to.  Used for tooltips and colouring the observations.
#' @param subset an optional vector specifying a subset of observations
#'   to be used for plotting. Useful when you have a large number of
#'   observations, you can specify a random subset.
#' @param labels the names of the variables (column names of \code{x}
#'   used by default).
#'
#' @import shiny
#'
#' @examples
#' data(iris)
#' \dontrun{
#' shinypairs(iris)
#' }
#'
#' @export
shinypairs = function(x, group=NULL, subset=NULL, labels = NULL){
  calls = match.call()
  shinyApp(
    ui=fluidPage(
      titlePanel(""),
      fluidRow(
        column(3,
               wellPanel(
                 uiOutput("varselect"),
                 uiOutput("facselect1"),
                 uiOutput("facselect2"),
                 radioButtons("theme", "Colour theme",
                              choices = c("Colour"= "colour",
                                          "Monochrome"="bw")),
                 #sliderInput("fontsize","Font size",12,min=6,max=24),
                 sliderInput("cex","Size of plotting symbol",3,min=1,max=10),
                 sliderInput("opacity","Opacity of plotting symbol",0.9,min=0,max=1),
                 sliderInput("width","Width and height",600,min=200,max=1200),
                 radioButtons("table_data_logical", label="Table of data?",
                              choices = c("No" = 0,
                                          "Yes" = 1)),
                 conditionalPanel("input.table_data_logical==1",
                                  selectInput(inputId="table_data_vars",label="Include all variables in table?",
                                              choices=c("No" = 0,
                                                        "Yes" = 1))
                 )
               ),
               wellPanel(
                 downloadButton("export",label="Download html file"),
                 br(),br(),
                 strong("Recreate using this code:"),
                 verbatimTextOutput("code")
               ),
               wellPanel(
                 icon("warning"),
                 tags$small("The pairsD3 package is under active development."),
                 tags$small("Report issues here: "),
                 HTML(paste("<a href=http://github.com/garthtarr/pairsD3/issues>")),
                 icon("github"),
                 HTML(paste("</a>"))
               )
        ),
        column(9,
               uiOutput("pairsplot"),
               br(),br(),
               dataTableOutput(outputId="outputTable")
        )
      )
    ),
    shinyServer(function(input, output) {

      output$varselect <- renderUI({
        cols = colnames(x)
        selectInput("choose_vars", "Select variables to plot",
                    choices=cols, selected=cols[1:3], multiple=T)
      })


      output$facselect1 <- renderUI({
        radioButtons("factor_var_logical", label="Is there a factor variable?",
                     choices = c("Yes" = 1,
                                 "No" = 0),
                     selected = selectedfac())
      })
      output$facselect2 <- renderUI({
        if(!is.null(calls[["group"]])){
          cols = c(deparse(calls[["group"]]),colnames(x))
        } else{
          cols = colnames(x)
        }
        conditionalPanel(
          condition = "input.factor_var_logical == 1",
          selectInput(inputId="factor_var",label="Factor variable:",
                      choices=cols,multiple=FALSE,selected = selectedfacvar())
        )
      })

      selectedfac = reactive({
        if(!is.null(calls[["group"]])) {
          return(1)
        } else return(0)
      })

      selectedfacvar = reactive({
        if(!is.null(calls[["group"]])) {
          return(NULL)
        } else { # tries to identify the most likely factor variable
          n.fac = function(x){length(levels(as.factor(x)))}
          nfacs = apply(x,2,n.fac)
          nfacs = nfacs[nfacs>1] # exclude any vars that are all
          return(names(which.min(nfacs))[1])
        }
      })

      groupvar = reactive({
        if(!is.null(input$factor_var_logical)){
          if(input$factor_var_logical==1){
            if(input$factor_var==deparse(calls[["group"]])){
              return(group)
            } else{
              return(x[,input$factor_var])
            }
          }
        }
      })

      output$pairsplot <- renderUI({
        pairsD3Output("pD3",width = input$width,height=input$width)
      })

      output$export = downloadHandler(
        filename = "pairsD3.html",
        content = function(file){
          savePairs(pairsD3(x,group=groupvar(), subset=subset, labels = labels,
                            theme = input$theme,
                            width=input$width,
                            opacity = input$opacity,
                            cex = input$cex),
                    file=file)
        }
      )

      output$pD3 <- renderPairsD3({
        pairsD3(x[,choices()],group=groupvar(), subset=subset, labels = labels,
                theme = input$theme, big=TRUE,
                opacity = input$opacity,
                cex = input$cex)
      })

      output$code = renderText({
        if(length(choices())==dim(x)[2]){
          paircall = paste("pairsD3(",deparse(calls[["x"]]),sep="")
        } else {
          paircall = paste("pairsD3(",deparse(calls[["x"]]),"[,c(",paste(match(choices(),names(x)),collapse=","),")]",sep="")
        }
        if(!is.null(calls[["group"]])){

        }
        if(!is.null(input$factor_var_logical)){
          if(input$factor_var_logical==1){
            if(input$factor_var==deparse(calls[["group"]])){
              paircall = paste(paircall,", group = ",deparse(calls[["group"]]),sep="")
            } else{
              paircall = paste(paircall,", group = ",deparse(calls[["x"]]),"[,",match(input$factor_var,names(x)),"]",sep="")
            }
          }
        }
        if(!is.null(calls[["subset"]])){
          paircall = paste(paircall,", subset = ",deparse(calls[["subset"]]),sep="")
        }
        if(!is.null(calls[["labels"]])){
          paircall = paste(paircall,", labels = ",deparse(calls[["labels"]]),sep="")
        }
        if(input$theme=="bw"){
          paircall = paste(paircall,", theme = 'bw'",sep="")
        }
        if(length(choices())>9){
          paircall = paste(paircall,", big = TRUE",sep="")
        }
        paircall = paste(paircall,", opacity = ", input$opacity,sep="")
        paircall = paste(paircall,", cex = ", input$cex,sep="")
        return(paste(paircall, ", width = ",input$width,")",sep=""))
      })

      choices<-reactive({
        input$choose_vars
      })

      output$outputTable = renderDataTable({
        data = x
        if(input$table_data_logical==1){
          displayDF <- as.matrix(data) # baseData$df #data sent to d3.js
          n=dim(displayDF)[1]
          dfFilter <- input$selectedobs[1:n] # passed from the web interface
          if (is.null(dfFilter)){
            # no selection has been made
            dfFilter = rep(TRUE,n)
          }
          displayDF <- as.data.frame(cbind(names=row.names(displayDF),
                                           displayDF))
          dfFilter[dfFilter==''] = TRUE
          dfFilter[dfFilter=='greyed'] = FALSE
          if(input$table_data_vars==0){
            return(as.matrix(displayDF[dfFilter == TRUE,choices(),drop=FALSE]))
          } else if(input$table_data_vars==1){
            return(as.matrix(displayDF[dfFilter == TRUE,,drop=FALSE]))
          }
        } else {
          return(NULL)
        }
      },
      options = list(dom = 't<lp>',pageLength = 20,
                     autoWidth = TRUE,
                     lengthMenu = list(c(20, 50, -1), c('20', '50', 'All')),
                     searching = FALSE)
      )

    })
  )}
